[SPARK-12155] [SPARK-12253] Fix executor OOM in unified memory management

andrewor14 opened the pull request:
**Problem.** In unified memory management, acquiring execution memory may lead to eviction of storage memory. However, the space freed from evicting cached blocks is distributed among all active tasks. Thus, an incorrect upper bound on the execution memory per task can cause the acquisition to fail, leading to OOM's and premature spills.

**Example.** Suppose total memory is 1000B, cached blocks occupy 900B, `spark.memory.storageFraction` is 0.4, and there are two active tasks. In this case, the cap on task execution memory is 100B / 2 = 50B. If task A tries to acquire 200B, it will evict 100B of storage but can only acquire 50B because of the incorrect cap. For another example, see this [regression test](https://github.com/andrewor14/spark/blob/fix-oom/core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala#L233) that I stole from @JoshRosen.

**Solution.** Fix the cap on task execution memory. It should take into account the space that could have been freed by storage in addition to the current amount of memory available to execution. In the example above, the correct cap should have been 600B / 2 = 300B.

This patch also guards against the race condition (SPARK-12253):
(1) Existing tasks collectively occupy all execution memory
(2) New task comes in and blocks while existing tasks spill
(3) After tasks finish spilling, another task jumps in and puts in a large block, stealing the freed memory
(4) New task still cannot acquire memory and goes back to sleep

Commit:
Pass in callbacks (gross)

Commit:
Notify all on task completion

Commit:
Rename silly method names + add detailed comments

andrewor14 commented:
@JoshRosen @davies

SparkQA commented:
**[Test build #47474 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47474/consoleFull)** for PR 10240 at commit [`cd0c680`](https://github.com/apache/spark/commit/cd0c680161e9c6f8044401ec1c2c3a4e83d4b6a1).

SparkQA commented:
**[Test build #47474 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47474/consoleFull)** for PR 10240 at commit [`cd0c680`](https://github.com/apache/spark/commit/cd0c680161e9c6f8044401ec1c2c3a4e83d4b6a1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `        logInfo(s\"HBase class not found $e\")`\n  * `        logDebug(\"HBase class not found\", e)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47474/
Test FAILed.

davies commented on core/src/main/scala/org/apache/spark/memory/ExecutionMemoryPool.scala:
```diff
@@ -91,23 +108,34 @@ private[memory] class ExecutionMemoryPool(
         // if we can't give it this much now, wait for other tasks to free up memory
         // (this happens if older tasks allocated lots of memory before N grew)
-        if (memoryFree >= math.min(maxToGrant, poolSize / (2 * numActiveTasks) - curMem)) {
+        if (memoryFree >= math.min(maxToGrant, poolSize / minMemoryPerTask)) {
```
`poolSize / minMemoryPerTask` should be `minMemoryPerTask - curMem`

andrewor14 replied:
oops good catch!!

davies commented on core/src/main/scala/org/apache/spark/memory/ExecutionMemoryPool.scala:
```diff
@@ -91,23 +108,34 @@ private[memory] class ExecutionMemoryPool(
       val toGrant = math.min(maxToGrant, memoryFree)
 
-      if (curMem < poolSize / (2 * numActiveTasks)) {
+      if (curMem < minMemoryPerTask) {
```
I'd like to change this block as this:
```
if (toGrant < numBytes && curMem + toGrant < minMemoryPerTask) {
  lock.wait()
} else {
  memoryForTask(taskAttemptId) += toGrant
  return toGrant
}
```

andrewor14 replied:
that looks fine, but is it completely equivalent to the old code? It's not 100% clear to me.

davies replied:
If not, the old code is not correct.

davies replied:
The current code is hard to understand, I can prove that it's the same with mine one.

andrewor14 replied:
yeah, I agree, though it's something we can always fix separately so we don't block the release. Let's defer the judgment to @JoshRosen.

andrewor14 replied:
I was able to prove this myself. I summarized my thoughts in this gist: https://gist.github.com/andrewor14/aea58796dd25d2ec9f20

That said, I would still prefer to do this separately since this PR is already passing tests. :)

Commit:
Fix unintended change

SparkQA commented:
**[Test build #47491 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47491/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).

SparkQA commented:
**[Test build #47491 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47491/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47491/
Test PASSed.

SparkQA commented:
**[Test build #2195 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2195/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).

SparkQA commented:
**[Test build #2195 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2195/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

andrewor14 commented:
ok, retest this please

andrewor14 commented:
Last commit actually passed tests last night.

SparkQA commented:
**[Test build #47528 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47528/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).

SparkQA commented:
**[Test build #47528 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47528/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47528/
Test PASSed.

Commit:
Simplify 1/2N code but preserve behavior

Per suggestion from @davies. For a detailed proof of why these
two are the same, see this gist:
https://gist.github.com/andrewor14/aea58796dd25d2ec9f20

andrewor14 commented:
retest this please

andrewor14 commented:
@davies please look at the final changes.

nongli commented on core/src/main/scala/org/apache/spark/memory/ExecutionMemoryPool.scala:
```diff
@@ -70,11 +70,28 @@ private[memory] class ExecutionMemoryPool(
+   * @param maybeGrowPool a callback that potentially grows the size of this pool. It takes in
+   *                      one parameter (Long) that represents the desired amount of memory by
+   *                      which this pool should be expanded.
+   * @param computeMaxPoolSize a callback that returns the maximum allowable size of this pool
```
If you take into account the memory that can be freed then isn't this a fixed value?

davies replied:
The actual used memory by storage could be changed, so it's not a fixed value

nongli replied:
Nvm, I see now.

andrewor14 replied:
no, because if storage memory used is below a certain mark (default 0.5 of max memory) then it cannot be evicted. In this case the max pool size depends on how much unevictable storage memory there is.

SparkQA commented:
**[Test build #47539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47539/consoleFull)** for PR 10240 at commit [`2929640`](https://github.com/apache/spark/commit/292964002cbbf695b621afd205269b0cb0198c83).

SparkQA commented:
**[Test build #2197 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2197/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).

SparkQA commented:
**[Test build #2196 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2196/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).

SparkQA commented:
**[Test build #2198 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2198/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).

davies commented:
LGTM

SparkQA commented:
**[Test build #2198 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2198/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `        logInfo(s\"HBase class not found $e\")`\n  * `        logDebug(\"HBase class not found\", e)`\n

SparkQA commented:
**[Test build #47539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47539/consoleFull)** for PR 10240 at commit [`2929640`](https://github.com/apache/spark/commit/292964002cbbf695b621afd205269b0cb0198c83).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47539/
Test PASSed.

SparkQA commented:
**[Test build #2197 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2197/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #2196 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2196/consoleFull)** for PR 10240 at commit [`d8be669`](https://github.com/apache/spark/commit/d8be66911d2abf3da46a25a54a7d80fd1eeebdfa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

andrewor14 commented:
Thanks, merging into master 1.6.
